### DISCUSSION TITLE
Unstable training

### DISCUSSION THREAD

**Discussion by WeberJulian** [coqui-ai/TTS#388] [General Q&A]
I'm training on french mailabs with GST, speaker embedding and mixed precision enabled and my training is very chaotic. (dev)

Here is my tensorboard: (the blue part correspond to a continue_training at 10k)
![image](https://user-images.githubusercontent.com/17219561/111612546-dd953f00-87dd-11eb-9c5f-0ffcd2c92bfa.png)

The test samples speak for themselves:

Here is 13.8k (very good):
https://sndup.net/6h5y

Here is 15.7k (hell):
https://sndup.net/7qmz

here is a link to my config:
https://pastebin.com/2iCygE62

**Comment by a-froghyar** [comment]
@WeberJulian which Attention mechanism are you using? Are you using T1 or T2? For me, training with T1 GST only yielded good results with the original attention, I think Graves has a bug with T1 - https://github.com/coqui-ai/TTS/issues/383 and the Dynamic Convolution yielded noise after 80k steps or so. Not sure about the dynamic conv but original attention definitely works, Graves just needs debugging.

**Reply by a-froghyar** [reply]
Plus, I'd actually let it train for longer, I usually only check the models after at least 20k steps

**Reply by WeberJulian** [reply]
I'm using Tacotron2 and original attention, I know it's a bit early on the training but  I think the behaviour is sufficiently weird to be posted here. Have you listened to the samples ?

**Reply by a-froghyar** [reply]
Yeah you're right, it probably went wrong somewhere, does this always happen?

**Reply by WeberJulian** [reply]
No first time, but it's also the first time I enabled mixed precision.

**Comment by erogol** [comment]
you can try ```noam_schedule: True``` to let model stabilize initially with lower learning rates. 

Also ```tb_model_param_stats:True``` to watch model layer stats on TensorBoard. It shows you if something is wrong with any of the layers.

**Reply by WeberJulian** [reply]
Thanks, I'll try that after trying without mixed precision

**Reply by lexkoro** [reply]
Sorry for going off-topic, but what do you think about the idea of adding also a LR value to the gradual training?

**Reply by WeberJulian** [reply]
What about ddc_r ? (we could put null when ddc is not enabled)

**Reply by erogol** [reply]
@SanjaESC Changing the LR does not make sense for me since the ultimate output length or the weights do not change.
But if you think it's useful then let's take it in a PR 

@WeberJulian I remember when I did that people asked for ```ddc: True``` or ```False``` :)

**Comment by WeberJulian** [comment]
So I tried with `mixed_precision: false` and it works for now 

![image](https://user-images.githubusercontent.com/17219561/111757457-a4bb9f80-889b-11eb-84aa-b415e96aeb05.png)

**Reply by WeberJulian** [reply]
Is there something that is inherently more unstable with mixed precision ?

**Reply by lexkoro** [reply]
Could you share a TrainIterStats screenshot from the previous run with the problems? I've had problems with the SSIM loss and mixed precision since it was generating NaN values.

**Reply by WeberJulian** [reply]
Sure, here you go:
![image](https://user-images.githubusercontent.com/17219561/111760328-dda94380-889e-11eb-8ac9-99eef433a249.png)
![image](https://user-images.githubusercontent.com/17219561/111760375-eb5ec900-889e-11eb-9361-d347ece9719e.png)

Should I set ignore outliers to false ?

**Reply by lexkoro** [reply]
Seems okay, at least I don't see any anomalies. 

I noticed you used a scale_stats. Maybe try without it and set "spec_gain" to 20. I've also tried scale_stats with a multi-speaker dataset once but the results were rather bad (quality wise). Some speakers had a screeching voice but others were normal.

**Reply by WeberJulian** [reply]
Thing is, I wanna be able to use the pretrained vocoders, so it's not really an option for me... I'd rather not use mixed precision and loose a bit of inference speed...

**Reply by lexkoro** [reply]
I might be wrong here but mixed precision (the config flag) is mostly related to training (using FP16) not inference speed.

**Reply by erogol** [reply]
it is about vram use.

**Reply by WeberJulian** [reply]
You mean specifically in the implementation of coqui-ai/TTS ? Or more broadly ?

**Reply by erogol** [reply]
I mean mixed precision is only for reducing vram use at training. You can use half precision model at inference but itndoenst always increase the speed .

**Reply by WeberJulian** [reply]
![image](https://user-images.githubusercontent.com/17219561/111769592-b146f480-88a9-11eb-9052-ddb90ce1acc5.png)

From what I read, it's indeed not obvious how to gain inference speed from mixed precision. But it seems possible through TensorRT so in solutions like ONNX

**Reply by WeberJulian** [reply]
This repo may be interesting: https://github.com/NVIDIA-AI-IOT/torch2trt

**Reply by WeberJulian** [reply]
@SanjaESC 
> Some speakers had a screeching voice but others were normal.

4 out of 5 speakers are great but one is completely broken. Is there a downside of spec_gain 20 vs scale_stats ? If no, wouldn't it be better to release a wavegrad model that works with those audio parameters to be compatible with more scenarios ?

**Reply by erogol** [reply]
In theory spec gain and scale stats are only about the normalization of  the data therefore they should not affect the final results with the right hyper parameters.  

So I am not convinced that spec-gain would make a big difference. Being said that I saw some papers that learn speaker specific normalization parameters for Batch Norm like layers . So maybe we can consider the same. 

Aside from the discussion, we can also just try and see how it works with spec-gain = 20 too :)

**Reply by erogol** [reply]
Is there any regularity among the broken speakers like gender voice type or so?

**Reply by WeberJulian** [reply]
I only have one broken speaker so I can't really generalize from that...  But here are the samples

Normal (male): https://sndup.net/3pwz
Normal (female): https://sndup.net/4mdf
Broken: https://sndup.net/6332

**Reply by WeberJulian** [reply]
@SanjaESC Is my broken sample representative of the issue you talked about ?

**Reply by lexkoro** [reply]
What exactly do you mean by broken? Personally I would say it sounds bit worse yeah, but without a reference audio it is hard to say. 
Also is this with wavegrad? If yes, does it also sound bad with GL?

**Reply by lexkoro** [reply]
One question regarding the speaker embeddings. 
Did you compute them using just the speaker-encoder? Or did you finetune the encoder first with the used dataset?

**Reply by WeberJulian** [reply]
I'll post samples of GT and GL later. Also I posted the config in the first post :) (for your speaker encoder question)
